mayastor/src/lvm/lvm.rs

```diff
@@ -0,0 +1,285 @@
+            Err(_) => "failed to execute vgs".to_string(),
+        };
+        return Err(Error::FailedExec{err: msg.to_string()})
+    }
+
+    let r: VGListReport = match serde_json::from_slice(output.stdout.as_slice()){
+        Ok(r) => r,
+        Err(e) => return Err(Error::FailedParsing{err: e.to_string()}),
+    };
+
+    let mut res: Vec<Pool> = vec![];
+
+    for v in r.report[0].vg.as_slice() {
+        let pool = match get_lvm_vg(v.vg_name.clone()).await {
+            Ok(p) => p,
+            Err(e) => return Err(e),
+        };
+        res.push(pool);
+    }
+
+    Ok(res)
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct VGsReport {
+    report: Vec<VG>
+}
+#[derive(Debug, Serialize, Deserialize)]
+struct VG {
+    vg: Vec<VGSize>
+}
+#[derive(Debug, Serialize, Deserialize)]
+struct VGSize {
+    vg_size: String,
+    vg_free: String
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct PVsReport {
+    report: Vec<PV>
+}
+#[derive(Debug, Serialize, Deserialize)]
+struct PV {
+    pv: Vec<VgPvMap>
+}
+#[derive(Debug, Serialize, Deserialize)]
+struct VgPvMap {
+    vg_name: String,
+    pv_name: String
+}
+
+pub async fn get_lvm_vg(name: String) -> Result<Pool, Error> {
+
+    let output = Command::new(VGS_COMMAND)
+        .arg(name.as_str())
+        .args(&["--options=vg_size,vg_free", "--units=b", "--noheadings", "--nosuffix", "--reportformat=json"])
+        .output()
+        .await
+        .expect("failed to execute vgs");
+    if !output.status.success() {
+        let msg = match String::from_utf8(output.stderr){
+            Ok(s) => s,
+            Err(_) => "failed to execute vgs".to_string(),
+        };
+        return Err(Error::FailedExec{err: msg.to_string()})
+    }
+
+    let r: VGsReport = match serde_json::from_slice(output.stdout.as_slice()){
+        Ok(r) => r,
+        Err(e) => return Err(Error::FailedParsing{err: e.to_string()}),
+    };
+
+    let capacity: u64 = match r.report[0].vg[0].vg_size.parse(){
+        Ok(c) => c,
+        Err(e) => return Err(Error::FailedParsing{err: e.to_string()})
+    };
+    let free: u64 = match r.report[0].vg[0].vg_free.parse(){
+        Ok(c) => c,
+        Err(e) => return Err(Error::FailedParsing{err: e.to_string()})
+    };
+
+    let output = Command::new(PVS_COMMAND)
+        .args(&["--options=vg_name,pv_name", "--noheadings", "--reportformat=json"])
+        .output()
+        .await
+       .expect("failed to execute vgs");
+    if !output.status.success() {
+        let msg = match String::from_utf8(output.stderr){
+            Ok(s) => s,
+            Err(_) => "failed to execute pvs".to_string(),
+        };
+        return Err(Error::FailedExec{err: msg.to_string()})
+    }
+
+    let r: PVsReport = match serde_json::from_slice(output.stdout.as_slice()){
+        Ok(r) => r,
+        Err(e) => return Err(Error::FailedParsing{err: e.to_string()}),
+    };
+
+    let mut disks: Vec<String> = vec![];
+
+    for p in r.report[0].pv.as_slice() {
```

Comment:
We might need checks against r.report not being empty to avoid panics.

Reply:
yeah. probably a good place to make use of `Option::ok_or`